File: src/storage/crypto.js
"use strict";

const crypto = require("node:crypto");

const META_KEY_KDF_SALT = "kdf_salt_b64";

function nowIso() {
  return new Date().toISOString();
}

function ensureKdfSalt(db) {
  const select = db.prepare("SELECT value FROM meta WHERE key = ?");
  const existing = select.get(META_KEY_KDF_SALT);
  if (existing && typeof existing.value === "string" && existing.value.trim()) {
    return existing.value.trim();
  }

  const salt = crypto.randomBytes(16).toString("base64");
  const insert = db.prepare("INSERT INTO meta (key, value) VALUES (?, ?)");
  insert.run(META_KEY_KDF_SALT, salt);
  return salt;
}

function deriveKey({ masterKey, saltB64 }) {
  const salt = Buffer.from(saltB64, "base64");
  return crypto.scryptSync(masterKey, salt, 32);
}

function encryptString({ plaintext, key }) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", key, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  const combined = Buffer.concat([iv, tag, ciphertext]);
  return `enc:v1:${combined.toString("base64")}`;
}

function decryptString({ value, key }) {
  if (typeof value !== "string") return value;
  if (!value.startsWith("enc:v1:")) return value;

  const b64 = value.slice("enc:v1:".length);
  const combined = Buffer.from(b64, "base64");
  const iv = combined.subarray(0, 12);
  const tag = combined.subarray(12, 28);
  const ciphertext = combined.subarray(28);

  const decipher = crypto.createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAuthTag(tag);
  const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  return plaintext.toString("utf8");
}

function createCryptoContext({ db, masterKey }) {
  const normalizedMasterKey = (masterKey || "").toString();
  const saltB64 = ensureKdfSalt(db);

  if (!normalizedMasterKey) {
    return {
      enabled: false,
      createdAt: nowIso(),
      encryptString: (v) => v,
      decryptString: (v) => {
        if (typeof v === "string" && v.startsWith("enc:v1:")) {
          throw new Error("HAZUKI_MASTER_KEY is required to decrypt stored secrets");
        }
        return v;
      },
    };
  }

  const key = deriveKey({ masterKey: normalizedMasterKey, saltB64 });
  return {
    enabled: true,
    createdAt: nowIso(),
    encryptString: (plaintext) => encryptString({ plaintext, key }),
    decryptString: (value) => decryptString({ value, key }),
  };
}

module.exports = { createCryptoContext };

File: go/internal/storage/crypto.go
package storage

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"database/sql"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"golang.org/x/crypto/scrypt"
)

const (
	metaKeyKdfSalt = "kdf_salt_b64"
	encPrefix      = "enc:v1:"
)

var ErrDecryptAuthFailed = errors.New("decrypt: message authentication failed")

type CryptoContext struct {
	Enabled bool
	key     []byte
}

func NewCryptoContext(db *sql.DB, masterKey string) (*CryptoContext, error) {
	saltB64, err := ensureKdfSalt(db)
	if err != nil {
		return nil, err
	}

	if strings.TrimSpace(masterKey) == "" {
		return &CryptoContext{Enabled: false}, nil
	}

	salt, err := base64.StdEncoding.DecodeString(saltB64)
	if err != nil {
		return nil, err
	}

	// Node crypto.scryptSync defaults: N=16384, r=8, p=1.
	key, err := scrypt.Key([]byte(masterKey), salt, 16384, 8, 1, 32)
	if err != nil {
		return nil, err
	}

	return &CryptoContext{Enabled: true, key: key}, nil
}

func NewCryptoContextFromSalt(masterKey, saltB64 string) (*CryptoContext, error) {
	if strings.TrimSpace(masterKey) == "" {
		return &CryptoContext{Enabled: false}, nil
	}
	salt, err := base64.StdEncoding.DecodeString(strings.TrimSpace(saltB64))
	if err != nil {
		return nil, err
	}

	// Node crypto.scryptSync defaults: N=16384, r=8, p=1.
	key, err := scrypt.Key([]byte(masterKey), salt, 16384, 8, 1, 32)
	if err != nil {
		return nil, err
	}
	return &CryptoContext{Enabled: true, key: key}, nil
}

func GetKdfSaltB64(db *sql.DB) (string, error) {
	return ensureKdfSalt(db)
}

func (c *CryptoContext) EncryptString(plaintext string) (string, error) {
	if !c.Enabled || strings.TrimSpace(plaintext) == "" {
		return plaintext, nil
	}
	if strings.HasPrefix(plaintext, encPrefix) {
		return plaintext, nil
	}

	block, err := aes.NewCipher(c.key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	iv := make([]byte, 12)
	if _, err := rand.Read(iv); err != nil {
		return "", err
	}

	ctWithTag := gcm.Seal(nil, iv, []byte(plaintext), nil)
	if len(ctWithTag) < gcm.Overhead() {
		return "", errors.New("encrypt: ciphertext too short")
	}
	tagLen := gcm.Overhead()
	ciphertext := ctWithTag[:len(ctWithTag)-tagLen]
	tag := ctWithTag[len(ctWithTag)-tagLen:]

	combined := make([]byte, 0, len(iv)+len(tag)+len(ciphertext))
	combined = append(combined, iv...)
	combined = append(combined, tag...)
	combined = append(combined, ciphertext...)

	return encPrefix + base64.StdEncoding.EncodeToString(combined), nil
}

func (c *CryptoContext) DecryptString(value string) (string, error) {
	if strings.TrimSpace(value) == "" {
		return value, nil
	}
	if !strings.HasPrefix(value, encPrefix) {
		return value, nil
	}
	if !c.Enabled {
		return "", errors.New("HAZUKI_MASTER_KEY is required to decrypt stored secrets")
	}

	b64 := strings.TrimPrefix(value, encPrefix)
	combined, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return "", err
	}
	if len(combined) < 12+16 {
		return "", errors.New("decrypt: payload too short")
	}
	iv := combined[:12]
	tag := combined[12:28]
	ciphertext := combined[28:]

	block, err := aes.NewCipher(c.key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	ctWithTag := make([]byte, 0, len(ciphertext)+len(tag))
	ctWithTag = append(ctWithTag, ciphertext...)
	ctWithTag = append(ctWithTag, tag...)

	plaintext, err := gcm.Open(nil, iv, ctWithTag, nil)
	if err != nil {
		if strings.Contains(err.Error(), "message authentication failed") {
			return "", errors.Join(ErrDecryptAuthFailed, err)
		}
		return "", err
	}
	return string(plaintext), nil
}

func ensureKdfSalt(db *sql.DB) (string, error) {
	var value string
	err := db.QueryRow("SELECT value FROM meta WHERE key = ?", metaKeyKdfSalt).Scan(&value)
	if err == nil && strings.TrimSpace(value) != "" {
		return strings.TrimSpace(value), nil
	}
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return "", err
	}

	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	salt := base64.StdEncoding.EncodeToString(b)

	_, err = db.Exec("INSERT INTO meta (key, value) VALUES (?, ?)", metaKeyKdfSalt, salt)
	if err != nil {
		return "", fmt.Errorf("insert kdf salt: %w", err)
	}
	return salt, nil
}
